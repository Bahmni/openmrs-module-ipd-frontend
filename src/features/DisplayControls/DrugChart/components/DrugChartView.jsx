import React, { useEffect, useState } from "react";
import PropTypes from "prop-types";
import { Button } from "carbon-components-react";
import { ChevronLeft16, ChevronRight16 } from "@carbon/icons-react";
import DrugChart from "./DrugChart";
import {
  fetchMedications,
  TransformDrugChartData,
  getDateTime,
  currentShiftHoursArray,
  getNextShiftDetails,
  getPreviousShiftDetails,
  SortDrugChartData,
  getDateFormatString,
} from "../utils/DrugChartUtils";
import { formatDate } from "../../../../utils/DateTimeUtils";
import data from "../../../../utils/config.json";
import { FormattedMessage } from "react-intl";
import "../styles/DrugChartView.scss";

const NoMedicationTaskMessage = (
  <FormattedMessage
    id={"NO_MEDICATION_DRUG_CHART_MESSAGE"}
    defaultMessage={"No Medication has been scheduled in this shift"}
  />
);

export default function DrugChartWrapper(props) {
  const { patientId } = props;
  const [drugChartData, setDrugChartData] = useState([]);
  const [isLoading, setLoading] = useState(true);
  const [date, updateDate] = useState(new Date());
  const [lastAction, updateLastActon] = useState("");
  const [startEndDates, updatedStartEndDates] = useState({
    startDate: new Date(),
    endDate: new Date(),
  });

  const { config: { drugChart = {} } = {} } = data;

  const [currentShiftArray, updateShiftArray] = useState(
    currentShiftHoursArray()
  );

  const callFetchMedications = async (startDateTime, endDateTime) => {
    const startDateTimeInSeconds = startDateTime / 1000;
    const endDateTimeInSeconds = endDateTime / 1000 - 60;
    try {
      const response = await fetchMedications(
        patientId,
        startDateTimeInSeconds,
        endDateTimeInSeconds
      );
      setDrugChartData(response.data);
    } catch (e) {
      return e;
    } finally {
      setLoading(false);
    }
  };

  useEffect(() => {
    const currentShift = currentShiftHoursArray();
    const startDateTime = getDateTime(new Date(), currentShift[0]);
    const endDateTime = getDateTime(
      new Date(),
      currentShift[currentShift.length - 1] + 1
    );
    updatedStartEndDates({ startDate: startDateTime, endDate: endDateTime });
    callFetchMedications(startDateTime, endDateTime);
  }, []);

  const sortedDrugChartData = SortDrugChartData(drugChartData);
  const transformedDrugChartData = TransformDrugChartData(sortedDrugChartData);

  const handlePrevious = () => {
    const firstHour = currentShiftArray[0];
    const lastHour = currentShiftArray[currentShiftArray.length - 1];
    if (lastHour < firstHour && lastAction === "N") {
      date.setDate(date.getDate() - 1);
    }
    const { startDateTime, endDateTime, nextDate } = getPreviousShiftDetails(
      currentShiftArray,
      drugChart.shiftHours,
      date
    );
    const previousShiftArray = currentShiftArray.map((hour) => {
      let updatedHour = hour - drugChart.shiftHours;
      updatedHour = updatedHour < 0 ? 24 + updatedHour : updatedHour;
      return updatedHour;
    });
    updateShiftArray(previousShiftArray);
    updateDate(nextDate);
    updateLastActon("P");
    setLoading(true);
    updatedStartEndDates({ startDate: startDateTime, endDate: endDateTime });
    callFetchMedications(startDateTime, endDateTime);
  };

  const handleNext = () => {
    const firstHour = currentShiftArray[0];
    const lastHour = currentShiftArray[currentShiftArray.length - 1];
    if (lastHour < firstHour && lastAction === "P") {
      date.setDate(date.getDate() + 1);
    }
    const { startDateTime, endDateTime, nextDate } = getNextShiftDetails(
      currentShiftArray,
      drugChart.shiftHours,
      date
    );
    const nextShiftArray = currentShiftArray.map(
      (hour) => (hour + drugChart.shiftHours) % 24
    );
    updateShiftArray(nextShiftArray);
    updateDate(nextDate);
    updateLastActon("N");
    setLoading(true);
    updatedStartEndDates({ startDate: startDateTime, endDate: endDateTime });
    callFetchMedications(startDateTime, endDateTime);
  };

  const handleCurrent = () => {
    const currentShift = currentShiftHoursArray();
    const nextDate = new Date();
    const startDateTime = getDateTime(new Date(), currentShift[0]);
    const endDateTime = getDateTime(
      new Date(),
      currentShift[currentShift.length - 1] + 1
    );
    updateShiftArray(currentShift);
    updateDate(nextDate);
    updateLastActon("");
    setLoading(true);
    updatedStartEndDates({ startDate: startDateTime, endDate: endDateTime });
    callFetchMedications(startDateTime, endDateTime);
  };
  const dateFormatString = getDateFormatString();
  return (
    <div className="drugchart-parent-container">
      <div className="drugchart-shift-header">
        <Button
          kind="tertiary"
          isExpressive
          size="small"
          onClick={handleCurrent}
          className="margin-right-10"
          data-testid="currentShift"
        >
          <FormattedMessage
            id={"CURRENT_SHIFT"}
            defaultMessage={"Current Shift"}
          />
        </Button>
        <Button
          renderIcon={ChevronLeft16}
          kind="tertiary"
          isExpressive
          hasIconOnly
          size="sm"
          onClick={handlePrevious}
          className="margin-right-6"
          data-testid="previousButton"
        />
        <Button
          renderIcon={ChevronRight16}
          kind="tertiary"
          isExpressive
          hasIconOnly
          size="sm"
          onClick={handleNext}
          className="margin-right-10"
          data-testid="nextButton"
        />
        <span>
          {`${formatDate(
            startEndDates.startDate,
            dateFormatString
          )} - ${formatDate(startEndDates.endDate, dateFormatString)}`}
        </span>
      </div>
      {isLoading ? (
        <div>Loading...</div>
      ) : drugChartData && drugChartData.length === 0 ? (
        <div className="no-nursing-tasks">{NoMedicationTaskMessage}</div>
      ) : (
        <DrugChart
          drugChartData={transformedDrugChartData}
          currentShiftArray={currentShiftArray}
          selectedDate={date}
        />
      )}
    </div>
  );
}
DrugChartWrapper.propTypes = {
  patientId: PropTypes.string.isRequired,
};
