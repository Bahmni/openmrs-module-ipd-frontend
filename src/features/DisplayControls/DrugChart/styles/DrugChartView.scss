.drugchart-parent-container {
  .drugchart-shift-header {
    display: flex;
    align-items: center;
    flex-wrap: wrap;
    margin: 10px 0;
  }

  .margin-right-6 {
    margin-right: 6px;
  }

  .margin-right-10 {
    margin-right: 10px;
  }

  .bx--btn--expressive {
    min-height: 32px;
    padding: 5px 10px;
    font-size: 14px;
    line-height: 18px;
    &:hover,
    &:focus {
      color: #fff;
    }
  }

  .bx--tooltip__trigger.bx--btn--icon-only--top.bx--tooltip--align-center::before {
    display: none;
  }

  .bx--tooltip__trigger.bx--btn--icon-only--top.bx--tooltip--align-center::after {
    display: none;
  }

  .bx--tooltip__trigger.bx--btn--icon-only--top.bx--tooltip--align-center
    .bx--assistive-text {
    display: none;
  }
}
