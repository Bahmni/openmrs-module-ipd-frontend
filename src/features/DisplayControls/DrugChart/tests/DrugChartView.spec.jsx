import React from "react";
import { render } from "@testing-library/react";
import DrugChartView from "../components/DrugChartView";
import { drugChartData } from "./DrugChartViewMockData";
import MockDate from "mockdate";

const mockUseFetchMedications = jest.fn();
const MockTooltipCarbon = jest.fn();

jest.mock("../hooks/useFetchMedications", () => ({
  useFetchMedications: () => mockUseFetchMedications(),
}));
jest.mock("bahmni-carbon-ui", () => {
  return {
    TooltipCarbon: (props) => {
      MockTooltipCarbon(props);
      return <div>TooltipCarbon</div>;
    },
  };
});

describe("DrugChartWrapper", () => {
  const mockDate = new Date(1466424490000);
  MockDate.set(mockDate);
  it("matches snapshot", () => {
    mockUseFetchMedications.mockReturnValue({
      isLoading: false,
      drugChartData,
    });
    const { container } = render(
      <DrugChartView patientId="testid" viewDate={mockDate} />
    );
    expect(container).toMatchSnapshot();
    MockDate.reset();
  });

  it("should render loading state when isLoading is true", () => {
    mockUseFetchMedications.mockReturnValue({
      isLoading: true,
      drugChartData: [],
    });
    const { container } = render(
      <DrugChartView patientId="test-id" viewDate={mockDate} />
    );
    expect(container).toMatchSnapshot();
  });

  it("should render no medication task message when drugChartData is empty", () => {
    mockUseFetchMedications.mockReturnValue({
      isLoading: false,
      drugChartData: [],
    });
    const { getByText } = render(
      <DrugChartView patientId="test-id" viewDate={mockDate} />
    );
    expect(
      getByText("No Medication has been scheduled for the patient yet")
    ).toBeTruthy();
  });
});
