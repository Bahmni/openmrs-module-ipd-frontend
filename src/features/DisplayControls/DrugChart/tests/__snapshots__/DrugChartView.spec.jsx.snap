// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DrugChartWrapper matches snapshot 1`] = `
<div>
  <div
    class="drug-chart-dashboard"
  >
    <div
      class="drug-chart"
      style="height: 198px;"
    >
      <div
        class="drug-chart-left-panel"
        data-testid="left-panel"
      >
        <div
          class="header"
        />
        <div>
          <table
            style="overflow: hidden;"
          >
            <tbody>
              <tr>
                <td>
                  <div
                    class="drug-name-container"
                  >
                    <div
                      class="drug-chart-drug-name"
                    >
                      Paracetamol 120 mg/5 mL Suspension (Liquid)
                    </div>
                    <div>
                      TooltipCarbon
                    </div>
                    <div
                      class="drug-list-drug-name-cell"
                    >
                      <div
                        class="bx--tag bx--tag--blue"
                        id="tag-2"
                      >
                        
                        <span
                          title="Rx"
                        >
                          Rx
                        </span>
                      </div>
                    </div>
                  </div>
                  <div
                    class="dosage"
                  >
                    <span>
                      25ml
                    </span>
                    <span>
                       - 
                      Oral
                    </span>
                    <span>
                       - 
                      3 Day(s)
                    </span>
                  </div>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
      <div
        class="drug-chart-content"
        data-testid="right-panel"
      >
        <div
          class="calendar-header"
        >
          <div
            style="display: flex;"
          >
            <div
              class="hour-header"
              data-testid="hour"
            >
              08
              :00
            </div>
            <div
              class="hour-header"
              data-testid="hour"
            >
              09
              :00
            </div>
            <div
              class="hour-header"
              data-testid="hour"
            >
              10
              :00
            </div>
            <div
              class="hour-header"
              data-testid="hour"
            >
              11
              :00
            </div>
            <div
              class="hour-header"
              data-testid="hour"
            >
              12
              :00
            </div>
            <div
              class="hour-header"
              data-testid="hour"
            >
              13
              :00
            </div>
            <div
              class="hour-header"
              data-testid="hour"
            >
              14
              :00
            </div>
            <div
              class="hour-header"
              data-testid="hour"
            >
              15
              :00
            </div>
          </div>
        </div>
        <table
          class="drug-chart-calendar"
        >
          <tbody>
            <tr>
              <div
                style="display: flex;"
              >
                <div
                  class="time-cell"
                >
                  <div
                    class=""
                    data-testid="left-icon"
                  />
                  <div
                    class=""
                    data-testid="right-icon"
                  />
                </div>
                <div
                  class="time-cell"
                >
                  <div
                    class=""
                    data-testid="left-icon"
                  />
                  <div
                    class=""
                    data-testid="right-icon"
                  />
                </div>
                <div
                  class="time-cell"
                >
                  <div
                    class=""
                    data-testid="left-icon"
                  />
                  <div
                    class=""
                    data-testid="right-icon"
                  />
                </div>
                <div
                  class="time-cell"
                >
                  <div
                    class=""
                    data-testid="left-icon"
                  />
                  <div
                    class=""
                    data-testid="right-icon"
                  />
                </div>
                <div
                  class="time-cell"
                >
                  <div
                    class="highligtedCell"
                    data-testid="left-icon"
                  />
                  <div
                    class=""
                    data-testid="right-icon"
                  />
                </div>
                <div
                  class="time-cell"
                >
                  <div
                    class=""
                    data-testid="left-icon"
                  />
                  <div
                    class=""
                    data-testid="right-icon"
                  />
                </div>
                <div
                  class="time-cell"
                >
                  <div
                    class=""
                    data-testid="left-icon"
                  />
                  <div
                    class=""
                    data-testid="right-icon"
                  />
                </div>
                <div
                  class="time-cell"
                >
                  <div
                    class=""
                    data-testid="left-icon"
                  />
                  <div
                    class=""
                    data-testid="right-icon"
                  />
                </div>
              </div>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
    <div
      class="drug-chart-legend"
    >
      <div>
        <test-file-stub />
        Pending
      </div>
      <div>
        <test-file-stub />
        Late
      </div>
      <div>
        <test-file-stub />
        Administered
      </div>
      <div>
        <test-file-stub />
        Administered Late
      </div>
      <div>
        <test-file-stub />
        Not Administered
      </div>
      <div>
        <test-file-stub />
        Stopped
      </div>
    </div>
  </div>
</div>
`;

exports[`DrugChartWrapper should render loading state when isLoading is true 1`] = `
<div>
  <div>
    Loading...
  </div>
</div>
`;
